### PR TITLE
[Fix]  Rendertexture and shadowmap platform adaptation

### PR DESCRIPTION
**Pull request by GengineJS**

### CODE REVIEW

**Comment by GengineJS** [comment]
https://github.com/cocos-creator/3d-tasks/issues/6645

**Review by minggo — APPROVED** [review]

File: editor/assets/chunks/cc-shadow-map-fs.chunk [hunk 1]
```diff
@@ -11,12 +12,16 @@ in highp vec4 v_shadowPos;
   #pragma builtin(global)
   layout(set = 0, binding = 5) uniform sampler2D cc_spotLightingMap;
 
+  void renderTextureFlip(inout vec2 uv) {
```
Comment (YunHsiao) [inline editor/assets/chunks/cc-shadow-map-fs.chunk]: This function encapsulation doesn't seem to be necessary

**Comment by minggo** [comment]
I think should add a documentation to describe how to write shader if using a texture drawn by GPU.

**Review by troublemaker52025 — APPROVED** [review]

File: editor/assets/chunks/cc-shadow-map-fs.chunk [hunk 2]
```diff
@@ -11,12 +12,16 @@ in highp vec4 v_shadowPos;
...
+    CC_HANDLE_NDC_SAMPLE_FLIP(uv);
+  }
+
   float CCGetShadowFactorX1 () {
     vec3 clipPos = v_shadowPos.xyz / v_shadowPos.w * 0.5 + 0.5;
     if (clipPos.x < 0.0 || clipPos.x > 1.0 ||
         clipPos.y < 0.0 || clipPos.y > 1.0 ||
         clipPos.z <-1.0 || clipPos.z > 1.0) { return 0.0; }
-
+    renderTextureFlip(clipPos.xy);
```
Comment (troublemaker52025) [inline editor/assets/chunks/cc-shadow-map-fs.chunk]: This action needs to be communicated to the user in the form of a document

File: editor/assets/chunks/common.chunk [hunk 3]
```diff
@@ -18,6 +18,8 @@
 #define STEP_FINT_128(v) (step(128.0, (v) + 0.5))
 #define MOD_FINT_128(v) (mod(float(int((v) + 0.5)), 128.0))
 
+#define CC_HANDLE_RT_SAMPLE_FLIP(uv) uv = floor(cc_cameraPos.w * 0.5) * 2.0 - 1.0 > 0.0 ? vec2(uv.x, 1.0 - uv.y) : uv
+#define CC_HANDLE_NDC_SAMPLE_FLIP(uv) uv = (floor(cc_cameraPos.w * 0.5) * 2.0 - 1.0) * (mod(cc_cameraPos.w, 2.0) * 2.0 - 1.0) > 0.0 ? uv : vec2(uv.x, 1.0 - uv.y)
```
Comment (troublemaker52025) [inline editor/assets/chunks/common.chunk]: Need to inform users when to use it

File: cocos/2d/framework/renderable-2d.ts [hunk 4]
```diff
@@ -522,6 +523,15 @@ export class Renderable2D extends RenderableComponent {
...
         }
+        if (this instanceof Sprite && this.spriteFrame && this.spriteFrame.texture instanceof RenderTexture) {
+            const defines = { SAMPLE_FROM_RT: true, ...mat.passes[0].defines };
+            const renderMat = new Material();
+            renderMat.initialize({
+                effectAsset: mat.effectAsset,
+                defines,
+            });
+            return renderMat;
+        }
```
Comment (troublemaker52025) [inline cocos/2d/framework/renderable-2d.ts:534]: Can you explain this part of the operation?

File: editor/assets/chunks/common.chunk [hunk 5]
```diff
@@ -18,6 +18,8 @@
 #define STEP_FINT_128(v) (step(128.0, (v) + 0.5))
 #define MOD_FINT_128(v) (mod(float(int((v) + 0.5)), 128.0))
 
+#define CC_HANDLE_RT_SAMPLE_FLIP(uv) uv = floor(cc_cameraPos.w * 0.5) * 2.0 - 1.0 > 0.0 ? vec2(uv.x, 1.0 - uv.y) : uv
```
Comment (YunHsiao) [inline editor/assets/chunks/common.chunk]: the branching condition could be simpified to `cc_cameraPos.w > 1.0`

File: editor/assets/chunks/common.chunk (hunk 3 above)
Comment (YunHsiao) [reply editor/assets/chunks/common.chunk]: `cc_cameraPos.w == 1.0` should work for this one

File: editor/assets/chunks/common.chunk (hunk 5 above)
Comment (YunHsiao) [reply editor/assets/chunks/common.chunk]: Add comments for this:
When sampling RTs, Y-flip is needed if the screen space sign Y is positive (pointing upwards)

File: editor/assets/chunks/common.chunk (hunk 3 above)
Comment (YunHsiao) [reply editor/assets/chunks/common.chunk]: This is not actually a common operation, but rather a special case for shadow map algorithms. So this should be put into shadow map headers.

File: cocos/core/pipeline/forward/forward-pipeline.ts [hunk 6]
```diff
@@ -345,6 +345,19 @@ export class ForwardPipeline extends RenderPipeline {
         }
     }
 
+    /**
+     * @en Map the value of [-1,1] in screenSpaceSignY and clipSpaceSignY to [0,1],and then move screenSpaceSignY one bit to the left and combine with clipSpaceSignY
+     * @zh 把screenSpaceSignY与clipSpaceSignY中[-1,1]的值映射为[0,1],通过screenSpaceSignY左移一位与clipSpaceSignY合并
```
Comment (YunHsiao) [inline cocos/core/pipeline/forward/forward-pipeline.ts]: keep the comments simple and stupid:

|combinedSignY|clipSpaceSignY|screenSpaceSignY|Backends|
|    :--:     |    :--:      |      :--:      |  :--:  |
|      0      |      -1      |      -1        | Vulkan |
|      1      |       1      |      -1        | Metal  |
|      2      |      -1      |       1        |        |
|      3      |       1      |       1        | GL variants   |

File: editor/assets/chunks/common.chunk (hunk 3 above)
Comment (troublemaker52025) [reply editor/assets/chunks/common.chunk]: sure

File: cocos/core/gfx/webgl/webgl-commands.ts [hunk 7]
```diff
@@ -1707,14 +1707,6 @@ export function WebGLCmdFuncBeginRenderPass (
         if (cache.glFramebuffer !== gpuFramebuffer.glFramebuffer) {
             gl.bindFramebuffer(gl.FRAMEBUFFER, gpuFramebuffer.glFramebuffer);
             cache.glFramebuffer = gpuFramebuffer.glFramebuffer;
-            // render targets are drawn with flipped-Y
-            const reverseCW = !!gpuFramebuffer.glFramebuffer;
-            if (reverseCW !== gfxStateCache.reverseCW) {
```
Comment (YunHsiao) [inline cocos/core/gfx/webgl/webgl-commands.ts:1712]: go ahead and delete `gfxStateCache.reverseCW` directly

File: cocos/core/gfx/device.ts [hunk 8]
```diff
@@ -352,11 +352,11 @@ export abstract class Device {
     }
 
     /**
-     * @en The sign of the UV space Y axis, positive if origin at upper-left.
-     * @zh UV 空间的 y 轴符号，原点在左上角时为正。
+     * @en The sign of the clip space Y axis, positive upward.
+     * @zh 裁剪空间的y轴符号，向上为正。
```
Comment (YunHsiao) [inline cocos/core/gfx/device.ts]: The wording for this comment should be consistent with `screenSpaceSignY`

File: editor/assets/chunks/common.chunk (hunk 3 above)
Comment (GengineJS) [reply editor/assets/chunks/common.chunk]: > `cc_cameraPos.w == 1.0` should work for this one

1 and 2 are different signs, 0 and 3 are the same signs,so can’t judge this way

File: editor/assets/chunks/common.chunk (hunk 5 above)
Comment (GengineJS) [reply editor/assets/chunks/common.chunk]: > the branching condition could be simpified to `cc_cameraPos.w > 1.0`

Yes it works

File: editor/assets/chunks/common.chunk (hunk 3 above)
Comment (YunHsiao) [reply editor/assets/chunks/common.chunk]: 2 doesn't correspond to any backends, so it's fine

**Review by troublemaker52025 — APPROVED** [review]

**Review by minggo — APPROVED** [review]

**Comment by GengineJS** [comment]
https://github.com/cocos-creator/engine-native/pull/3360

**Review by YunHsiao — APPROVED** [review]

**Pull request merged**